controllers/agentmachine_controller.go

```diff
@@ -180,10 +173,44 @@ func (r *AgentMachineReconciler) handleDeletionFinalizer(ctx context.Context, lo
 	return nil, nil
 }
 
-func (r *AgentMachineReconciler) findAgent(ctx context.Context, log logrus.FieldLogger, agentMachine *capiproviderv1alpha1.AgentMachine) (ctrl.Result, error) {
+func (r *AgentMachineReconciler) getAgentCluster(ctx context.Context, log logrus.FieldLogger, machine *clusterv1.Machine) (*capiproviderv1alpha1.AgentCluster, error) {
+	cluster, err := clusterutil.GetClusterFromMetadata(ctx, r.Client, machine.ObjectMeta)
+	if err != nil {
+		log.Info("Machine is missing cluster label or cluster does not exist")
+		return nil, err
+	}
+
+	agentClusterRef := types.NamespacedName{Name: cluster.Spec.InfrastructureRef.Name, Namespace: cluster.Spec.InfrastructureRef.Namespace}
+	agentCluster := &capiproviderv1alpha1.AgentCluster{}
+	if err := r.Get(ctx, agentClusterRef, agentCluster); err != nil {
+		log.WithError(err).Errorf("Failed to get agentCluster %s", agentClusterRef)
+		return nil, err
+	}
+
+	return agentCluster, nil
+}
+
+func (r *AgentMachineReconciler) findAgent(ctx context.Context, log logrus.FieldLogger, agentMachine *capiproviderv1alpha1.AgentMachine,
+	clusterDeploymentRef capiproviderv1alpha1.ClusterDeploymentReference, machineConfigPool string,
+	ignitionTokenSecretRef *aiv1beta1.IgnitionEndpointTokenReference) (ctrl.Result, error) {
+
+	foundAgent, err := r.findDanglingAgent(ctx, log, agentMachine)
+	if err != nil {
+		log.WithError(err).Error("failed to list dangling agents")
+		return ctrl.Result{RequeueAfter: defaultRequeueAfterOnError}, err
+	}
+	if foundAgent != nil {
+		err = r.updateAgentMachineWithFoundAgent(ctx, log, agentMachine, foundAgent)
+		if err != nil {
+			log.WithError(err).Error("failed to update AgentMachine with found dangling agent")
+			return ctrl.Result{RequeueAfter: defaultRequeueAfterOnError}, err
+		} else {
+			return ctrl.Result{Requeue: true}, nil
+		}
```

Comment:
I think you can refactor the function so this logic happens once (at the end).
It doesn't matter if the agent was dangling or not.